### PR TITLE
fix(python): Don't allow passing missing data to generalized ufuncs

### LINKED ISSUE #14811: generalized ufuncs that return arrays are broken in a variety of cases
This is a follow-up to #14748.

### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import numpy as np
import polars as pl
from numba import guvectorize, float64


# ISSUE 1: Missing data silently corrupts the results
@guvectorize([(float64[:], float64[:])], "(m)->(m)")
def mean(arr, result):
    value = arr.mean()
    for i in range(len(arr)):
        result[i] = value


series = pl.Series([2, 2, 2, None], dtype=pl.Float64)
print("== ISSUE 1 ==")
print("Mean should be 2, but we get 1.5:", mean(series))
print()

# ISSUE 2: Output arrays with different size than input array are not handled by Polars:
@guvectorize([(float64[:], float64[:], float64[:])], "(n),(m)->(m)")
def different_output_size(arr, arr2, result):
    total = arr.mean()
    for i in range(len(arr2)):
        result[i] = total


arr = np.array([1, 2, 3], dtype=np.float64)
arr2 = np.arange(0, 1_000_000, dtype=np.float64)

assert len(different_output_size(arr, arr2)) == len(arr2)
assert len(different_output_size(arr2, arr)) == len(arr)

print("== ISSUE 2 ==")
series = pl.Series(arr, dtype=pl.Float64)
series2 = pl.Series(arr2, dtype=pl.Float64)
assert len(different_output_size(series, series2)) == len(series2)
assert len(different_output_size(series2, series)) == len(series)
```

### Log output

```shell
== ISSUE 1 ==
Mean should be 2, but we get 1.5: shape: (4,)
Series: '' [f64]
[
        1.5
        1.5
        1.5
        null
]

== ISSUE 2 ==
thread '<unnamed>' panicked at crates/polars-arrow/src/compute/boolean_kleene.rs:105:5:
assertion `left == right` failed: lhs and rhs must have the same length
  left: 3
 right: 1000000
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "/home/itamarst/devel/polars/py-polars/../guvectorize2.py", line 36, in <module>
    assert len(different_output_size(series, series2)) == len(series2)
  File "/home/itamarst/devel/polars/.venv/lib/python3.10/site-packages/numba/np/ufunc/gufunc.py", line 172, in __call__
    return self.ufunc(*args, **kwargs)
  File "/home/itamarst/devel/polars/py-polars/polars/series/series.py", line 1434, in __array_ufunc__
    validity_mask &= arg.is_not_null()
  File "/home/itamarst/devel/polars/py-polars/polars/series/series.py", line 642, in __and__
    return self._from_pyseries(self._s.bitand(other._s))
pyo3_runtime.PanicException: assertion `left == right` failed: lhs and rhs must have the same length
  left: 3
 right: 1000000
```

### Issue description

The current `ufunc` support essentially assumes just `ufunc`, and can't handle the variety of edge cases generalized `ufunc`s add:

1. **MISSING-DATA:** Missing data can give unexpected results, since processing is of the whole array. This is documented but many people won't notice those docs and just get bad data unknowningly. Even if you do read the docs, it can happen by accident.
2. **DIFFERENT-OUTPUT-SIZE**: The current code assumes the output array is the same as all input arrays, which is not a requirement for generalized `ufunc`s. I suspect there is an edge case where different output size can result in memory corruption...

There are also potentially edge cases in the signature syntax that can be rejected early rather than late, with a more informative error message, e.g. multi-dimensional output maybe.

### Expected behavior

This is an open question. Here is what I would expect:

1. **MISSING-DATA:** Raise an exception, e.g. `ValueError("There is missing data in in your Series, which can result in generalized ufuncs giving wrong results")`.
2. **DIFFERENT-OUTPUT-SIZE:** DTRT and return a `Series` of the correct size. The reason identical input and output size is assumed right now is (a) because it's always true for normal `ufunc` and (b) to deal with missing data (c) it's a nice optimization. However, the code could be changed to allow the `ufunc` to allocate the output array, and then that data copied into a `Series`.

Note that that both these proposals are for generalized `ufunc` only. Regular `ufunc` would continue as normal.

| Type of `ufunc` | Behavior with missing data | Who allocates memory |
|------------------------|-----------------------------------------|-----------------------------------|
| Normal               | Accepted, since the `ufunc` runs on individual values anyway | Polars, as an optimization |
| Generalized       | Rejected with exception, for correctness                                      | The `ufunc` |

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.13
Index type:           UInt32
Platform:             Linux-6.6.10-76060610-generic-x86_64-with-glibc2.35
Python:               3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
```

</details>

### PR DESCRIPTION
**Pull request by itamarst**
Fixes #14811 

This is a more constrained PR, just addressing one specific issue: passing missing data to generalized ufuncs can silently lead to incorrect results.

Generalized ufuncs are a NumPy feature (https://numpy.org/neps/nep-0020-gufunc-signature-enhancement.html), but the easiest way to implement them is using Numba, which is what the tests do.

### CODE REVIEW

File: py-polars/tests/unit/interop/numpy/test_ufunc_series.py [hunk 1]
```diff
@@ -119,3 +121,31 @@ def test_numpy_string_array() -> None:
...
+    If a pl.Series is missing data, using a generalized ufunc is not allowed.
+
+    While this particular example isn't necessarily a semantic issue, consider
+    a mean() function running on integers: it will give wrong results if the
+    input is missing data, since NumPy has no way to model missing slots.  In
+    the general case, we can't assume the function will handle missing data
+    correctly.
+    """
+    s_float = pl.Series("f", [1.0, 2.0, 3.0, None], dtype=pl.Float64)
+    with pytest.raises(pl.ComputeError):
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/interop/numpy/test_ufunc_series.py:150]: perhaps use `match=`, to verify it's not raising because of some unrelated reason?

**Review by MarcoGorelli — COMMENTED** [review]
Implementation looks good to me, just got some quite minor comments

File: py-polars/tests/unit/interop/numpy/_numba.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+"""
+Infrastructure for testing Numba.
+
+Numba releases often lag for a few months after Python releases, so we don't
+want Numba to be a blocker for Python 3.X support. So this minimally emulates
+the Numba module, while allowing for the fact that Numba may not be installed.
```
[the remark this answers is not in the document]
Comment (itamarst) [reply py-polars/tests/unit/interop/numpy/_numba.py:6]: I guess I could do that if I defined the Numba function inside the test?

**Review by MarcoGorelli — APPROVED** [review]
Thanks @itamarst !

**Comment by itamarst** [comment]
Thank you @MarcoGorelli. If you have commit access you'll have to merge it too, since I can't.

**Comment by MarcoGorelli** [comment]
yup - just leaving it open another day in case anyone has comments, then I'll ship it

outside of generalized ufuncs, there's no other case where missing values are a concern, right?

**Comment by itamarst** [comment]
For normal value-by-value ufuncs, if you have memory that is e.g. an invalid float64 then that could potentially be a problem? You're potentially feeding random garbage to the function for missing data.

**Pull request merged**